interp/src/flatten/structures/environment/wave.rs

```diff
@@ -0,0 +1,220 @@
+                            *signal_id,
+                            value.to_bit_str().as_bytes(),
+                        )?;
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+impl Drop for WaveWriter {
+    fn drop(&mut self) {
+        let writer = std::mem::take(&mut self.writer);
+        if let Some(writer) = writer {
+            writer.finish().unwrap();
+        }
+    }
+}
+
+enum Todo {
+    /// instance name, instance id
+    OpenScope(String, GlobalCellIdx),
+    CloseScope,
+}
+
+type PortToSignalMap = AuxillaryMap<GlobalPortIdx, Option<FstSignalId>>;
+
+fn declare_signals<
+    C: AsRef<Context> + Clone,
+    W: std::io::Write + std::io::Seek,
+>(
+    out: &mut FstHeaderWriter<W>,
+    env: &Environment<C>,
+) -> Result<PortToSignalMap> {
+    let mut port_map: PortToSignalMap = PortToSignalMap::new();
+    let root_idx = GlobalCellIdx::new(0);
+    let root_comp = &env.ctx.as_ref().secondary[env.ctx().entry_point];
+    let root_name = env.ctx().lookup_name(root_comp.name).clone();
+    let mut todo = vec![Todo::OpenScope(root_name, root_idx)];
+    while let Some(component) = todo.pop() {
+        match component {
+            Todo::OpenScope(name, id) => {
+                todo.push(Todo::CloseScope);
+                declare_component(
+                    out,
+                    env,
+                    &mut todo,
+                    &mut port_map,
+                    id,
+                    name,
+                )?;
+            }
+            Todo::CloseScope => {
+                out.up_scope()?;
+            }
+        }
+    }
+    Ok(port_map)
+}
+
+fn declare_component<
+    C: AsRef<Context> + Clone,
+    W: std::io::Write + std::io::Seek,
+>(
+    out: &mut FstHeaderWriter<W>,
+    env: &Environment<C>,
+    todo: &mut Vec<Todo>,
+    port_map: &mut PortToSignalMap,
+    component_cell_idx: GlobalCellIdx,
+    instance_name: String,
+) -> Result<()> {
+    let instance = env.cells[component_cell_idx].as_comp().unwrap();
+    let component = &env.ctx.as_ref().secondary[instance.comp_id];
+
+    let component_name = &env.ctx().lookup_name(component.name);
+    out.scope(instance_name, component_name, FstScopeType::Module)?;
+
+    // component ports
+    declare_ports(
+        out,
+        env,
+        component
+            .signature_in
+            .iter()
+            .chain(component.signature_out.iter())
+            .map(|local| {
+                (
+                    component.port_offset_map[local],
+                    &instance.index_bases + local,
+                )
+            }),
+        port_map,
+    )?;
+
+    // child components
+    for (local_offset, cell) in component.cell_offset_map.iter() {
+        let cell = env
+            .ctx
+            .as_ref()
+            .secondary
+            .local_cell_defs
+            .get(*cell)
+            .unwrap();
+        let cell_idx = &instance.index_bases + local_offset;
```

Comment:
Can make it a little more compact with `.ctx()` and direct indexing since we
know the idx is valid
```suggestion
        let cell = &env.ctx().secondary.local_cell_defs[*cell];
```